x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutResponseHandler.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.authc.saml;
+
+import org.elasticsearch.common.unit.TimeValue;
+import org.opensaml.saml.saml2.core.LogoutResponse;
+import org.w3c.dom.Element;
+
+import java.time.Clock;
+import java.util.Collection;
+
+import static org.elasticsearch.xpack.security.authc.saml.SamlUtils.samlException;
+
+public class SamlLogoutResponseHandler extends SamlResponseHandler {
+
+    private static final String LOGOUT_RESPONSE_TAG_NAME = "LogoutResponse";
+
+    public SamlLogoutResponseHandler(
+        Clock clock, IdpConfiguration idp, SpConfiguration sp, TimeValue maxSkew) {
+        super(clock, idp, sp, maxSkew);
+    }
+
+    public void handle(byte[] payload, Collection<String> allowedSamlRequestIds) {
+        final Element root = parseSamlMessage(payload);
+        if (LOGOUT_RESPONSE_TAG_NAME.equals(root.getLocalName()) && SAML_NAMESPACE.equals(root.getNamespaceURI())) {
+            final LogoutResponse logoutResponse = buildXmlObject(root, LogoutResponse.class);
+            if (logoutResponse == null) {
+                throw samlException("Cannot convert element {} into LogoutResponse object", root);
+            }
+            if (logoutResponse.isSigned()) {
+                validateSignature(logoutResponse.getSignature());
+            }
+            checkInResponseTo(logoutResponse, allowedSamlRequestIds);
```

Comment:
This requires Kibana to know the `_ID` of the logout request crafted by ES and sent to the idP. The current `SamlLogoutResponse` does not give this information in a way that Kibana can easily access. 

We could either change `SamlLogoutResponse` to include a separate field of `_ID` similar to what we do with `SamlPrepareAuthenticationResponse`. Or we could simply skip this check?

Reply:
The spec ([4.4.4.1](https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf)) doesn't talk about validating the `InResponseTo` parameter, but I've seen other libraries / impementation do that. In all honesty SAML SLO is such a wild wild west, everyone does pretty much what they feel like :/ 

Given that it's not dictated by the specification and that our API has a binary response of "The Logout Response is fine" vs "the Logout Response is _not_ fine" which cannot be abused somehow ( ie. open redirect etc) I think it would be fine to not validate

Reply:
It's straightforward to add a new `requestId` field to the `SamlLogoutResponse` ES produces (via the `/_security/saml/logout` API). Given it is a change to an existing API, we may need to communicate to the Kibana team earlier on. @legrego Could you please comment on this? The `requestId` value is then required to be passed back to ES from Kibana to verify the `LogoutResponse` from the idP.

Reply:
> Given it is a change to an existing API, we may need to communicate to the Kibana team earlier on

In general, adding additional fields to our responses is not considered to be a breaking change so I believe Kibana will be fine with it even if we don't sync merging support for this. 

> The requestId value is then required to be passed back to ES from Kibana to verify the LogoutResponse from the idP.

Kibana does that already for the request ID of the authentication response and send this back to ES in the `_security/saml/authenticate` API so I believe it would be fine to do here too.

Reply:
Thanks @jkakavas I'll make the change accordingly.

Reply:
> In general, adding additional fields to our responses is not considered to be a breaking change so I believe Kibana will be fine with it even if we don't sync merging support for this.

That's correct.

> Kibana does that already for the request ID of the authentication response and send this back to ES in the _security/saml/authenticate API so I believe it would be fine to do here too.

The main difference here is that request ID returned from `_security/saml/prepare` is stored in the cookie. But during logout flow we used to clear cookie when user is redirected to IdP with `LogoutRequest`. Having said that, we can change that, we'll just need to carefully fix all the remaining places that still treat _any_ cookie as a sign of authenticated user.

Reply:
Since the specification _doesn't_ explicitly call this out *AND* the outcome is in any case a `no-op`, I would be ok with not validating this. Especially since it sounds possible that the risk of introducing a bug on Kibana's side by changing this behavior is not minimal. On the other hand, treating "any cookie as a sign of authenticated user." is something worth fixing regardless of that change so if this is an extra driver for that change, let's do it !